lib/Makefile

```diff
@@ -223,42 +222,78 @@ lib-all: all
 
 all: lib
 
-$(ZSTD_STATLIB_DIR)/libzstd.a: ARFLAGS = rcs
-$(ZSTD_STATLIB_DIR)/libzstd.a: | $(ZSTD_STATLIB_DIR)
-$(ZSTD_STATLIB_DIR)/libzstd.a: $(ZSTD_STATLIB_OBJ)
+ifeq ($(BUILD_DIR),0)
+# determine a BUILD_DIR
+
+.PHONY: libzstd.a  # not the actual recipe
+libzstd.a:
+	$(MAKE) --no-print-directory $@ \
+        BUILD_DIR=obj/$(HASH_DIR) \
+        CPPFLAGS="$(CPPFLAGS)"
+
+else
+# $(BUILD_DIR) is defined
+
+ZSTD_STATLIB_DIR := $(BUILD_DIR)/static
+ZSTD_STATLIB := $(ZSTD_STATLIB_DIR)/libzstd.a
+ZSTD_STATLIB_OBJ := $(addprefix $(ZSTD_STATLIB_DIR)/,$(ZSTD_LOCAL_OBJ))
+$(ZSTD_STATLIB): ARFLAGS = rcs
+$(ZSTD_STATLIB): | $(ZSTD_STATLIB_DIR)
+$(ZSTD_STATLIB): $(ZSTD_STATLIB_OBJ)
 	@echo compiling static library
-	$(Q)$(AR) $(ARFLAGS) $@ $^
+	$(AR) $(ARFLAGS) $@ $^
 
 .PHONY: libzstd.a  # must be run every time
-libzstd.a: $(ZSTD_STATLIB_DIR)/libzstd.a
-	$(Q)ln -sf $< $@
+libzstd.a: $(ZSTD_STATLIB)
+	ln -sf $< $@
+
+endif
 
 ifneq (,$(filter Windows%,$(TARGET_SYSTEM)))
 
 LIBZSTD = dll\libzstd.dll
-$(LIBZSTD):
 $(LIBZSTD): $(ZSTD_FILES)
 	@echo compiling dynamic library $(LIBVER)
 	$(CC) $(FLAGS) -DZSTD_DLL_EXPORT=1 -Wl,--out-implib,dll\libzstd.dll.a -shared $^ -o $@
 
 else
 
 LIBZSTD = libzstd.$(SHARED_EXT_VER)
-$(ZSTD_DYNLIB_DIR)/$(LIBZSTD): CFLAGS += -fPIC
-$(ZSTD_DYNLIB_DIR)/$(LIBZSTD): LDFLAGS += -shared -fvisibility=hidden
-$(ZSTD_DYNLIB_DIR)/$(LIBZSTD): | $(ZSTD_DYNLIB_DIR)
-$(ZSTD_DYNLIB_DIR)/$(LIBZSTD): $(ZSTD_DYNLIB_OBJ)
+$(LIBZSTD): CFLAGS += -fPIC
+$(LIBZSTD): LDFLAGS += -shared -fvisibility=hidden
+
+ifeq ($(BUILD_DIR),0)
```

Comment:
Can you add a comment somewhere explaining the 2-stage methodology, and why it is needed.

It isn't obvious to me why it is necessary.